### PR TITLE
NAS-128399 / 24.10 / More nfs failover fixes

### PR DESCRIPTION
**Pull request by mgrimesix**
### Problem
NFS was not configured to properly support HA function.  NFS holds client state information in a database that, by default, is located at /var/lib/nfs.  To properly support failover we need that to be a shared directory.  For TrueNAS that is the system dataset.

### Fix
Create and NFS entry in the system dataset.  Generate the necessary directory structure and to support upgrade we also initialize it with data from /var/lib/nfs.
In systemd we bind the NFS client tracking daemon to nfs-server to allow the transfer of the client tracking database.
Update the nfs configuration file with the path to the nfs data on the system dataset: `/var/db/system/nfs`

### Other related items in the PR

1. Updated the CI tests
2. Minor flake8 cleanup and refactoring

### CODE REVIEW

File: src/middlewared/middlewared/plugins/nfs.py
Comment (anodos325) [inline src/middlewared/middlewared/plugins/nfs.py:113]: I don't think this should be async.

File: src/middlewared/middlewared/plugins/nfs.py
Comment (anodos325) [inline src/middlewared/middlewared/plugins/nfs.py:30]: Rather than using pwd and grp here, you should inside method call `user.get_builtin_user_id`. See `set_etc_file_perms()` in etc.py for example.

**Review by yocalebo — CHANGES_REQUESTED** [review]

File: src/middlewared/middlewared/plugins/nfs.py
Comment (mgrimesix) [reply src/middlewared/middlewared/plugins/nfs.py:152]: This setting is mostly done for completeness.  
It's a proc file in support of _legacy_ handling of NFSv4 client state.  It does not exist if `nfs-server` is not running.

At 'boot' this is hard-coded to `/var/lib/nfs/v4recovery`
`static char user_recovery_dirname[PATH_MAX] = "/var/lib/nfs/v4recovery";`

The routine `setup_directories` is called once at nfs-server start, which does introduce a window for TOCTOU.

The 'x' mode is for creation.  I'm not going to be creating this file.  If it doesn't exist, then no harm.
It's the same for 'w'.  I think 'r+' _is_ the proper choice here.
https://docs.python.org/3/library/functions.html#open

File: src/middlewared/middlewared/plugins/nfs.py
Comment (mgrimesix) [reply src/middlewared/middlewared/plugins/nfs.py]: Made suggested change

File: src/middlewared/middlewared/plugins/nfs.py
Comment (mgrimesix) [reply src/middlewared/middlewared/plugins/nfs.py]: Implemented usage of get_builtin_user_id as a post-create fixup.

File: src/middlewared/middlewared/plugins/nfs.py
Comment (mgrimesix) [reply src/middlewared/middlewared/plugins/nfs.py]: Made suggested change.

File: src/middlewared/middlewared/plugins/nfs.py
Comment (mgrimesix) [reply src/middlewared/middlewared/plugins/nfs.py]: Converted to a scandir solution.

File: src/middlewared/middlewared/plugins/nfs.py
Comment (anodos325) [inline src/middlewared/middlewared/plugins/nfs.py:32]: This should just be a path, right? `platform` was something we were doing when there was mixed FreeBSD / Linux consumers of the source.

Comment (mgrimesix) [reply src/middlewared/middlewared/plugins/nfs.py:32]: path is a more sensible name.

File: src/middlewared/middlewared/plugins/nfs.py
Comment (yocalebo) [inline src/middlewared/middlewared/plugins/nfs.py:163]: Sorry, I most be missing something obvious so bear with me 😄  I'm still not following why you're only catching `FileNotFoundError`? Previously, you checked for files existence and then wrote to it catching any exception that may come up during the write I/O.

Now you're only catching if the file doesn't exist and then logging an error but you will crash here if writing to it fails.

What I'm getting at is, does it matter if this file exists or not?

Comment (mgrimesix) [reply src/middlewared/middlewared/plugins/nfs.py:163]: I don't think it matters if the file exists.   
This file was used in support of an early solution for maintaining client state.  It was supplanted with the database managed by `nfsdcld`.  However, the code supporting the old mechanism was not removed.  Since it's not yet been removed, I figure it's best to 'manage' it.
Side note:  More recently, `nfsdcld` itself has been replaced with `nfsdcltrk`, but Debian still uses `nfsdcld`.   Currently, all three are in Debian, but to my knowledge only `nfsdcld` is really 'active'.
I'll change to trap on any error.  It's more safe.

File: src/middlewared/middlewared/plugins/nfs.py
Comment (yocalebo) [inline src/middlewared/middlewared/plugins/nfs.py:117]: let's move the `name_to_id_conversion` inner method to its own `@private` method in this class. Using inner methods actually incurs a non-trivial amount of overhead. You can keep the same name if you want. I also see that if `usergrp[0]` fails to be resolved, then we won't try and resolve the group (usergrp[1]). I'm not sure if that's by design but I believe we should make the method a bit more robust and explicit. I also see that the log message is talking about "applying ownership" but it's doing none of the sorts. Looks like it's just resolving the name to a uid/gid 😄 I picture the method to look something like:
```
@private
def name_to_id_converstion(name, name_type='user'):
    if any((not isinstance(name, str), isinstance(name, int))):
        # it's not a string (NoneType, float, w/e) or it's an int
        # so there is nothing to do
        return name

    if name_type = 'user':
        method = 'user.get_builtin_user_id'
    elif name_type == 'group':
        method = 'group.get_builtin_group_id'
    else:
        self.logger.error('Unexpected name_type (%r)', name_type)
        return name
    try:
        return self.middleware.call_sync(method, name)
    except Exception as e:
        if hasattr(e, 'errno') and e.errno == errno.ENOENT:
            self.logger.error('Failed to resolve builtin %s %r', name_type, name)
        else:
             self.logger.error('Unexpected error resolving builtin %s %r, name_type, name, exc_info=True)
        return name

File: src/middlewared/middlewared/plugins/nfs.py
Comment (yocalebo) [inline src/middlewared/middlewared/plugins/nfs.py:147]: This for loop is prone to a few different race conditions. It's also using the "file handle" instead of opening the file/directory and then using the fd based syscalls. I think we need to simplify this to something simple.
     
Something like:
```
for i in NFSServicePathInfo:
    uid = self.name_to_id_conversion(i.owner()['uid'], name_type='user')
    gid = self.name_to_id_conversion(i.owner()['gid'], name_type='group')
    path = i.path()
    if i.is_dir():
        open_flags = os.O_RDWR | os.O_DIRECTORY
        os.makedirs(path, exist_ok=True)
    else:
        open_flags = os.O_RDWR

    try:
        with open(os.open(path, flags=open_flags)) as f:
            os.fchmod(f.fileno(), i.mode())
            os.fchown(f.fileno(), uid, gid)
    except IsADirectoryError:
        if not i.is_dir():
            # the path is expected to be a file but it's a directory
            self.logger.error('Expected %r to be a file but instead found a directory', path)
        else:
            self.logger.error('Unexpected failure updating file %r', path, exc_info=True)
    except NotADirectoryError:
        if i.is_dir():
            # the path is expected to be a directory but it's a file
            self.logger.error('Expected %r to be a directory but instead found a file', path)
        else:
            self.logger.error('Unexpected failure updating directory %r', path, exc_info=True)
    except Exception:
        self.logger.error('Unexpected failure initializing %r', path, exc_info=True)

**Review by yocalebo — CHANGES_REQUESTED** [review]
Just a few remaining items 👍

File: src/middlewared/middlewared/plugins/nfs.py
Comment (mgrimesix) [reply src/middlewared/middlewared/plugins/nfs.py:131]: Wait... the point of the scandir check was to allow the copy operation _**only**_ if /var/db/system/nfs is empty.

**Review by yocalebo — APPROVED** [review]
Fantastic, great job! Glad to see this general issue get resolved, finally. Be sure and wait for @anodos325 before merge 👍

File: src/middlewared/middlewared/plugins/nfs.py
Comment (yocalebo) [inline src/middlewared/middlewared/plugins/nfs.py:106]: woops, a typo in the method name 😄 
```suggestion
    def name_to_id_converstion(self, name, name_type='user'):
```

File: src/middlewared/middlewared/plugins/nfs.py
Comment (mgrimesix) [reply src/middlewared/middlewared/plugins/nfs.py]: Implemented the simplified solution.

**Review by yocalebo — APPROVED** [review]
nice!

**Review by anodos325 — APPROVED** [review]

**Pull request merged**
